mxfusion/modules/gp_modules/dgp.py

```diff
@@ -0,0 +1,430 @@
+from ...inference.variational import VariationalInference
+from ...models import Model, Posterior
+from ...modules.module import Module
+from ...runtime.distributions.multivariate_normal import MultivariateNormalRuntime
+from ...util.customop import make_diagonal
+
+
+def gaussian_variational_expectation(F, y, variance, f_mean, f_var):
+    """
+    :param F: mx.nd or mx.sym
+    :param y: observed y
+    :param variance: likelihood variance
+    :param f_mean: prediction mean
+    :param f_var: prediction variance
+    """
+    y = F.transpose(y, (0, 2, 1))
+    return -0.5 * np.log(2 * np.pi) - 0.5 * F.log(variance) \
+           - 0.5 * (F.square(y - f_mean) + f_var) / variance
+
+
+def predict_from_layer(F, inputs, layer, variables, posterior, model):
+    """
+    Computes p(f) for a particular layer given the x locations to predict at
+
+    :param F: mx net mode
+    :param inputs: Input locations to predict at
+    :param layer: zero-based layer index to predict from
+    :param variables:
+    :param posterior:
+    :param model:
+    :return:
+    """
+    q_u = get_q_u(F, layer, posterior, variables)
+
+    # Get p(f|u) object
+    z = variables[getattr(model, 'inducing_inputs_' + str(layer))]
+    conditional_gp = getattr(model, 'F_' + str(layer))
+    kern = getattr(model, 'kern_' + str(layer))
+    kern_params = kern.fetch_parameters(variables)
+    mu = variables[getattr(posterior, 'qU_mean_' + str(layer))]
+    rv_shape = mu.shape[1:]
+
+    samples, z = arrays_as_samples(mx.nd, [inputs, z])
+
+    q_u = q_u.broadcast_to_n_samples(samples.shape[0])
+    return conditional_gp.factor.marginalise_conditional_distribution(samples, z, q_u, rv_shape[0], 1e-8, True,
+                                                                      **kern_params)
+
+
+def propagate_samples_through_layers(F, variables, model, posterior, n_layers, n_samples, final_layer=None,
+                                     start_layer=None):
+    """
+    Propagates samples from start_layer to final_layer
+
+    :param F: mx.nd or mx.sym
+    :param variables: variables dictionary
+    :param model: model graph
+    :param posterior: posterior graph
+    :param n_layers: number of layers
+    :param n_samples: number of samples to propagate through layers
+    :param final_layer: Zero-based index of last layer to propagate samples through
+    :param start_layer: Zero-based index of first layer to propagate samples through
+    :return: (samples, mean, variance)
+    """
+
+    if final_layer is None:
+        final_layer = n_layers
+
+    if start_layer is None:
+        start_layer = 0
+
+    x = variables[model.X]
+
+    samples = x
+
+    for layer in range(start_layer, final_layer):
+        # Predict from layer
+        p_f = predict_from_layer(F, samples, layer, variables, posterior, model)
+
+        # Draw samples
+        samples = p_f.draw_samples(n_samples)
+        samples = F.transpose(samples, (0, 2, 1))
```

Comment:
Avoid this transpose here, do at beginning

Reply:
I'm struggling to get rid of these transposes. The fundamental problem is that kernels expect (n_samples, n_data, n_dims) inputs and the multivariate normal expects (n_samples, n_outputs, n_data, n_data). Going between layers n_outputs = n_dims, so some tranpose is necessary (I think?) without redoing the kernel functions (which I don't think we want).

Reply:
I don't see the kernel calculation will be problem. The kernel input does not have the dimension about n_outputs is because different output dimensions share the same prior covariance matrix. You can expand the resulting dimensionality of a kernel function, which is (n_samples, n_data, n_data) with an additional dimension (n_samples, 1, n_data, n_data).

Reply:
So I'll explain a bit more clearly:
Output from layer 1 is a `MultivariateNormalRuntime` with dimensions `(n_samples, n_outputs, n_data)`
We sample from this distribution to get inputs to layer 2 of shape `(n_samples, n_outputs, n_data)`
Now `n_outputs_layer_1 == n_inputs_layer_2`
Can the kernels take in as input an array of shape `(n_samples, n_inputs, n_data)`?

Reply:
I see. Let me think about it.

Reply:
I had a look at adding a "tranpose_output" flag to `MultivariateNormalRuntime` but couldn't work out how to do it within the linear algebra routines. The problem is we do a matrix matrix multiply with two arrays of shape:
 `(n_samples, n_outputs, n_inputs, n_inputs)`  x `(n_samples, n_outputs, n_inputs, 1)`

The transpose flags can get us output arrays of shape `(n_samples, n_outputs, 1, n_inputs)` instead of the normal `(n_samples, n_outputs, n_inputs, 1)`, but we actually want an output of shape `(n_samples, n_inputs, n_outputs)` which I can't figure out how to do. Any ideas?

Reply:
You are right. In this case, we probably cannot avoid transpose. Then, let's just leave it there.